### ISSUE TITLE
perl_eval_{sv,pv} does not trap syntax errors

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2380]
Migrated from [rt.perl.org#3719](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3719) (status was 'resolved')

Searchable as RT3719$

**Comment by p5pRT** [comment]
### From @gisle

#### Created by gisle@aas.no



I have this application \(actually Python\) that embeds perl and calls
perl\_eval\_pv\(\) to pass code to it\.  One problem with this is that
perl\_eval\_pv\(code\, FALSE\) does not set ERRSV \(aka $@&#8203;\) on syntax
errors\.

This is a little example program showing this behaviour&#8203;:

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\#include \<EXTERN\.h>
\#include \<perl\.h>

int main\(int argc\, char\*\*argv\, char\*\*env\)
\{
&nbsp;  int i;
&nbsp;  char \*embedding\[\] = \{""\, "\-e"\, "0"\};
&nbsp;  PerlInterpreter \*my\_perl = perl\_alloc\(\);
&nbsp;  perl\_construct\(my\_perl\);
&nbsp;  perl\_parse\(my\_perl\, NULL\, 3\, embedding\, NULL\);
&nbsp;  perl\_run\(my\_perl\);

&nbsp;  perl\_eval\_pv\("sub \{ "\, TRUE\);
&nbsp;  /\*\*\* the program ought to abort here \*\*\*/

&nbsp;  if \(SvTRUE\(ERRSV\)\)
&nbsp;       PerlIO\_printf\(Perl\_debug\_log\, "XXX %s"\, SvPV\_nolen\(ERRSV\)\);
&nbsp;  perl\_eval\_pv\("1/0"\, TRUE\);
&nbsp;  if \(SvTRUE\(ERRSV\)\)
&nbsp;       PerlIO\_printf\(Perl\_debug\_log\, "YYY %s"\, SvPV\_nolen\(ERRSV\)\);
&nbsp;  /\*\*\* it aborts here \*\*\*/

&nbsp;  printf\("done\\n"\);
\}

/\*
&nbsp;  Compile with&#8203;: cc check\-eval\-syntax\-error\.c \`perl \-MExtUtils&#8203;::Embed \-e ccopts \-e ldopts\`
 \*/
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

This program prints&#8203;:

&nbsp;   $ \./a\.out
&nbsp;   Illegal division by zero at \(eval 2\) line 2\.

instead of something with "syntax error in \(eval 1\)"\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by gisle at Mon May 22 17:07:41 CEST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.5, archname=i686-linux
    uname='linux eik 2.2.5 #1 mon sep 13 14:53:01 cest 1999 i686 unknown '
    config_args='-Dprefix=/local/perl/5.6.0_priv -Doptimize=-g -ders'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-g', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /local/perl/5.6.0_priv/lib/5.6.0/i686-linux
    /local/perl/5.6.0_priv/lib/5.6.0
    /local/perl/5.6.0_priv/lib/site_perl/5.6.0/i686-linux
    /local/perl/5.6.0_priv/lib/site_perl/5.6.0
    /local/perl/5.6.0_priv/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/gisle
    LANG=POSIX
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=...
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @gannett-ggreer

On Wed Aug 16 05&#8203;:18&#8203;:16 2000\, gaas wrote&#8203;:
> I have this application \(actually Python\) that embeds perl and calls
> perl\_eval\_pv\(\) to pass code to it\.  One problem with this is that
> perl\_eval\_pv\(code\, FALSE\) does not set ERRSV \(aka $@&#8203;\) on syntax
> errors\.
> 
> This is a little example program showing this behaviour&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \#include \<EXTERN\.h>
> \#include \<perl\.h>
> 
> int main\(int argc\, char\*\*argv\, char\*\*env\)
> \{
>   int i;
>   char \*embedding\[\] = \{""\, "\-e"\, "0"\};
>   PerlInterpreter \*my\_perl = perl\_alloc\(\);
>   perl\_construct\(my\_perl\);
>   perl\_parse\(my\_perl\, NULL\, 3\, embedding\, NULL\);
>   perl\_run\(my\_perl\);
> 
>   perl\_eval\_pv\("sub \{ "\, TRUE\);
>   /\*\*\* the program ought to abort here \*\*\*/
> 
>   if \(SvTRUE\(ERRSV\)\)
>        PerlIO\_printf\(Perl\_debug\_log\, "XXX %s"\, SvPV\_nolen\(ERRSV\)\);
>   perl\_eval\_pv\("1/0"\, TRUE\);
>   if \(SvTRUE\(ERRSV\)\)
>        PerlIO\_printf\(Perl\_debug\_log\, "YYY %s"\, SvPV\_nolen\(ERRSV\)\);
>   /\*\*\* it aborts here \*\*\*/
> 
>   printf\("done\\n"\);
> \}
> 
> /\*
>   Compile with&#8203;: cc check\-eval\-syntax\-error\.c \`perl \-MExtUtils&#8203;::Embed
> \-e ccopts \-e ldopts\`
>  \*/
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> This program prints&#8203;:
> 
>    $ \./a\.out
>    Illegal division by zero at \(eval 2\) line 2\.
> 
> instead of something with "syntax error in \(eval 1\)"\.

My embed\-fu is weak\, but it appears to be a simple fix\.

http&#8203;://m\-l\.org/~perl/misc/0001\-Fix\-for\-perl\-3719\-Preserve\-for\-perl\_eval\_pv\-while\-em\.patch

http&#8203;://github\.com/greerga/perl/commit/8ba06c38ccf1582a01143760851f9396c16b78f5

or attached\.

\-\- 
George Greer

**Comment by p5pRT** [comment]
### From @gannett-ggreer

<details><summary>0001-Fix-for-perl-3719-Preserve-for-perl_eval_pv-while-em.patch</summary>

```diff
From 8ba06c38ccf1582a01143760851f9396c16b78f5 Mon Sep 17 00:00:00 2001
From: George Greer <perl@greerga.m-l.org>
Date: Sun, 25 Jul 2010 21:08:40 -0400
Subject: [PATCH] Fix for [perl #3719]: Preserve $@ for perl_eval_pv() while embedding.

Perl_eval_pv() wraps around Perl_eval_sv() but Perl_eval_sv() throws
away the error before Perl_eval_pv() can see it.
---
 perl.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/perl.c b/perl.c
index d52d79f..13076b1 100644
--- a/perl.c
+++ b/perl.c
@@ -2765,7 +2765,7 @@ Perl_eval_pv(pTHX_ const char *p, I32 croak_on_error)
 
     PERL_ARGS_ASSERT_EVAL_PV;
 
-    eval_sv(sv, G_SCALAR);
+    eval_sv(sv, G_SCALAR | G_KEEPERR);
     SvREFCNT_dec(sv);
 
     SPAGAIN;
-- 
1.7.0.4


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jul 25 18&#8203;:18&#8203;:26 2010\, greerga wrote&#8203;:
> My embed\-fu is weak\, 

So is mine\.

> but it appears to be a simple fix\.
> 
> http&#8203;://m\-l\.org/~perl/misc/0001\-Fix\-for\-perl\-3719\-Preserve\-for\-
> perl\_eval\_pv\-while\-em\.patch
> 
>
http&#8203;://github\.com/greerga/perl/commit/8ba06c38ccf1582a01143760851f9396c16b78f5
> 
> or attached\.

Is anyone else able to review this?

**Comment by p5pRT** [comment]
### From @Leont

On Mon\, Jul 26\, 2010 at 3&#8203;:18 AM\, George Greer via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> My embed\-fu is weak\, but it appears to be a simple fix\.
>
> http&#8203;://m\-l\.org/~perl/misc/0001\-Fix\-for\-perl\-3719\-Preserve\-for\-perl\_eval\_pv\-while\-em\.patch
>
> http&#8203;://github\.com/greerga/perl/commit/8ba06c38ccf1582a01143760851f9396c16b78f5
>
> or attached\.

Disclaimer&#8203;: I have embed\-fu\, but not so much internals\-fu\.

The G\_KEEPERR flag seems to affect eval in at least 3 places
1\) in doeval\, right before compilation is tried\.
2\) in pp\_leave\, as the very last thing is does\. I think this is only run on
3\) in eval\_sv\, right after a successful eval\.

I suspect this patch has a wider impact than intended\. I don't really
know if that's really a problem though\.

Also\, the fact that this G\_KEEPERR behavior of eval\_sv is not
documented bothers me\.

Leon

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Sep 27\, 2010 at 12&#8203;:44&#8203;:16AM \+0200\, Leon Timmermans wrote&#8203;:
> On Mon\, Jul 26\, 2010 at 3&#8203;:18 AM\, George Greer via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > My embed\-fu is weak\, but it appears to be a simple fix\.
> >
> > http&#8203;://m\-l\.org/~perl/misc/0001\-Fix\-for\-perl\-3719\-Preserve\-for\-perl\_eval\_pv\-while\-em\.patch
> >
> > http&#8203;://github\.com/greerga/perl/commit/8ba06c38ccf1582a01143760851f9396c16b78f5
> >
> > or attached\.
> 
> Disclaimer&#8203;: I have embed\-fu\, but not so much internals\-fu\.
> 
> The G\_KEEPERR flag seems to affect eval in at least 3 places
> 1\) in doeval\, right before compilation is tried\.
> 2\) in pp\_leave\, as the very last thing is does\. I think this is only run on
> 3\) in eval\_sv\, right after a successful eval\.
> 
> I suspect this patch has a wider impact than intended\. I don't really
> know if that's really a problem though\.

The patch is wrong\, unfortunately\. It only "works" in some circumstances
because it exploits another bug whereby G\_KEEPERR still sets $@&#8203; sometimes\,

In particular&#8203;: under the original fix\, if $@&#8203; wasn't null when calling
eval\_pv\(pv\,TRUE\)\, it would croak\, even if pv compiled and ran okay\.

I've fixed the original bug with commit

&nbsp;    4aca2f62efca883199d7975f34b7fb876c280366

and fixed the G\_KEEPERR issue with

&nbsp;    6b2fb389897a68db92c38c22f79403607f0da303

> Also\, the fact that this G\_KEEPERR behavior of eval\_sv is not
> documented bothers me\.

I've fixed that with

&nbsp;    be064c4a1f59651142e99788731ae1d164c19537

\-\- 
Thank God I'm an atheist\.\.\.\.\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
